regcomp.c

```diff
@@ -5715,6 +5715,16 @@ S_study_chunk(pTHX_
                     delta += (minnext + deltanext) * maxcount
                              - minnext * mincount;
                 }
+
+                if (data && data->flags & SCF_SEEN_ACCEPT) {
+                    if (flags & SCF_DO_SUBSTR) {
```

Comment:
No, not as far as I can tell. Although while verifying this I found a problem caused by different code. The start class stuff should be fine as long as the min-length logic is correct. I will push an additional commit to deal with that.

Reply:
So the reason (*ACCEPT) doesn't affect stclass (eg ANYOF style) in the optimizer in the same way it does substrings is because *ACCEPT affects the minimum length string that the pattern might match, but it doesn't affect the things it might start with. The only real effect it might have on startclass would be to allow it to accept the empty string. Put another way, (*ACCEPT) means that things to its right cant be mandatory, but they can still be something that starts a match. The only case where this is not the case is when ACCEPT would cause the pattern to match the empty and in that case there shouldnt be a startclass at all, and it should be handled via the minlen logic. This is why a couple of the later patches in this PR are there, to ensure that minlen is set properly.  BTW, those patches wouldn't exist if you hadn't asked this question so thanks!

Consider a case like this (using charclasses here to prevent this being turned into a mandatory substring question).

/[aA]*[bB]/

it should have a stclass of ANYOF[aAbB] and have a minlen of 1.

But now if we insert an (*ACCEPT) in the middle:

/[aA]*(*ACCEPT)[bB]/

it has a minlen of 0, and will match any string and the optimizer already understands that if the pattern can match the empty string then there is no startclass at all.

Change the * to a +, and the ACCEPT doesn't affect the stclass at all, it will be the same with or without the ACCEPT, any ANYOF[aA].

Consider a case like / [aA]* ( [bB] | [cC](*ACCEPT) | [dD]*) [eE] /x, this should have a stclass of ANYOF[aAbBcCdDeE]. The ACCEPT doesn't affect the fact that 'e' might be a legit start character for a matching string.

With mandatory substrings however the case is different. If I have /a(?:b+(*ACCEPT))?c/ then the fixed string should be 'ab', c will not be part of a successful match. Compare with /ab+c/, where we have two mandatory strings to choose from 'ab' and 'bc'. 

With /a* ( b | c | d ) e/ the mandatory substring would be 'e', but if i inject an accept into the alternation like this: /a* (b | c (*ACCEPT) | d ) e/ then 'e' is now not mandatory, if it goes through the 'c' branch then 'e' will not be part of a match.

Reply:
Thanks, I'd overlooked the fact that minlen=0 makes stclass irrelevant: I guess that was implied by your previous comment, but I didn't pick up on it.

(I'd also recommend putting backticks round code fragments: one of your examples above is being parsed as URL markup in github's web interface, because it matches `[...](...)`.)